service/machinelearning/internal/customizations/predictendpoint.go

```diff
@@ -0,0 +1,61 @@
+package customizations
+
+import (
+	"context"
+	"fmt"
+	"github.com/awslabs/smithy-go"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+	"net/url"
+)
+
+type fetchPredictEndpointFunc func(interface{}) (*string, error)
+
+// AddPredictEndpointMiddleware adds the middleware required to set the endpoint
+// based on Predict's PredictEndpoint input member.
+func AddPredictEndpointMiddleware(stack *middleware.Stack, endpoint fetchPredictEndpointFunc) {
+	stack.Serialize.Insert(&predictEndpointMiddleware{}, "ResolveEndpoint", middleware.After)
```

Comment:
At some point we need to figure out a strategy for these middleware IDs and be constants somehow. Potentially with middleware slots, this may be easier. https://github.com/awslabs/smithy-go/issues/156